File: clients/go-sdk/client/event_types.go
package client

import (
	"context"
	"encoding/json"
	"fmt"
)

// EventTypeListResponse — GET /api/event-types returns `{ items: [...] }`.
type EventTypeListResponse struct {
	Items []EventTypeResponse `json:"items"`
}

// CreateEventTypeRequest is the body for POST /api/event-types.
type CreateEventTypeRequest struct {
	// Code follows {app}:{domain}:{aggregate}:{event}.
	Code string `json:"code"`
	// Human-readable name.
	Name string `json:"name"`
	// Optional description.
	Description string `json:"description,omitempty"`
	// Optional initial JSON schema. Use json.RawMessage to keep the
	// shape opaque end-to-end.
	Schema json.RawMessage `json:"schema,omitempty"`
	// Client ID for multi-tenant scoping.
	ClientID string `json:"clientId,omitempty"`
}

// UpdateEventTypeRequest is the body for PUT /api/event-types/{id}.
type UpdateEventTypeRequest struct {
	Name        *string `json:"name,omitempty"`
	Description *string `json:"description,omitempty"`
}

// AddSchemaVersionRequest is the body for POST /api/event-types/{id}/versions.
type AddSchemaVersionRequest struct {
	Schema json.RawMessage `json:"schema"`
}

// EventTypeResponse is the platform's event-type representation.
type EventTypeResponse struct {
	ID            string                `json:"id"`
	Code          string                `json:"code"`
	Name          string                `json:"name"`
	Description   string                `json:"description,omitempty"`
	Status        string                `json:"status"`
	Application   string                `json:"application,omitempty"`
	Subdomain     string                `json:"subdomain,omitempty"`
	Aggregate     string                `json:"aggregate,omitempty"`
	EventName     string                `json:"event,omitempty"`
	SpecVersions  []SpecVersionResponse `json:"specVersions,omitempty"`
	CreatedAt     string                `json:"createdAt"`
	UpdatedAt     string                `json:"updatedAt"`
}

// SpecVersionResponse is one schema version on an event type.
type SpecVersionResponse struct {
	Version string          `json:"version"`
	Status  string          `json:"status"`
	Schema  json.RawMessage `json:"schema,omitempty"`
}

// SyncEventTypesRequest is the body for the per-app sync endpoint.
type SyncEventTypesRequest struct {
	EventTypes []CreateEventTypeRequest `json:"eventTypes"`
}

// EventTypesResource is the accessor for /api/event-types/*.
// Construct via FlowCatalystClient.EventTypes.
type EventTypesResource struct {
	c *FlowCatalystClient
}

// Create — POST /api/event-types.
func (r *EventTypesResource) Create(ctx context.Context, req *CreateEventTypeRequest) (*EventTypeResponse, error) {
	var out EventTypeResponse
	if err := r.c.Post(ctx, "/api/event-types", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get — GET /api/event-types/{id}.
func (r *EventTypesResource) Get(ctx context.Context, id string) (*EventTypeResponse, error) {
	var out EventTypeResponse
	if err := r.c.Get(ctx, "/api/event-types/"+id, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetByCode — GET /api/event-types/by-code/{code}.
func (r *EventTypesResource) GetByCode(ctx context.Context, code string) (*EventTypeResponse, error) {
	var out EventTypeResponse
	if err := r.c.Get(ctx, "/api/event-types/by-code/"+code, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List — GET /api/event-types?application=&status=&clientId=.
func (r *EventTypesResource) List(ctx context.Context, application, status, clientID string) (*EventTypeListResponse, error) {
	q := EncodeQuery("application", application, "status", status, "clientId", clientID)
	var out EventTypeListResponse
	if err := r.c.Get(ctx, "/api/event-types"+q, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update — PUT /api/event-types/{id}.
func (r *EventTypesResource) Update(ctx context.Context, id string, req *UpdateEventTypeRequest) (*EventTypeResponse, error) {
	var out EventTypeResponse
	if err := r.c.Put(ctx, "/api/event-types/"+id, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// AddSchemaVersion — POST /api/event-types/{id}/versions.
func (r *EventTypesResource) AddSchemaVersion(ctx context.Context, id string, req *AddSchemaVersionRequest) (*EventTypeResponse, error) {
	var out EventTypeResponse
	if err := r.c.Post(ctx, "/api/event-types/"+id+"/versions", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Archive — DELETE /api/event-types/{id}. The platform soft-archives:
// the row is retained with status flipped to ARCHIVED. Named Archive
// (not Delete) to make that explicit.
func (r *EventTypesResource) Archive(ctx context.Context, id string) error {
	return r.c.Delete(ctx, "/api/event-types/"+id, nil)
}

// Sync — POST /api/applications/{appCode}/event-types/sync. Declarative
// reconciliation: the request's eventTypes list becomes the desired
// state. With removeUnlisted=true the server archives any existing
// event types not in the request.
func (r *EventTypesResource) Sync(ctx context.Context, appCode string, req *SyncEventTypesRequest, removeUnlisted bool) (*SyncResult, error) {
	q := ""
	if removeUnlisted {
		q = "?removeUnlisted=true"
	}
	var out SyncResult
	path := fmt.Sprintf("/api/applications/%s/event-types/sync%s", appCode, q)
	if err := r.c.Post(ctx, path, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: crates/fc-sdk/src/client/event_types.rs
//! Event Type management operations.

use super::{ClientError, FlowCatalystClient};
use serde::{Deserialize, Serialize};

/// List of event types returned by `GET /api/event-types`.
///
/// The platform uses `{ items: [...] }` — not `{ data, total }`. There is
/// no separate total count.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EventTypeListResponse {
    pub items: Vec<EventTypeResponse>,
}

/// Request to create an event type.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CreateEventTypeRequest {
    /// Code in format `{app}:{domain}:{aggregate}:{event}` (e.g., "orders:fulfillment:shipment:shipped")
    pub code: String,
    /// Human-readable name
    pub name: String,
    /// Optional description
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Optional initial JSON schema
    #[serde(skip_serializing_if = "Option::is_none")]
    pub schema: Option<serde_json::Value>,
    /// Client ID for multi-tenant scoping
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_id: Option<String>,
}

/// Request to update an event type.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateEventTypeRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// Request to add a schema version to an event type.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddSchemaVersionRequest {
    /// JSON schema for this version
    pub schema: serde_json::Value,
}

/// Event type response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EventTypeResponse {
    pub id: String,
    pub code: String,
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    pub status: String,
    #[serde(default)]
    pub application: String,
    #[serde(default)]
    pub subdomain: String,
    #[serde(default)]
    pub aggregate: String,
    #[serde(default, rename = "event")]
    pub event_name: String,
    #[serde(default)]
    pub spec_versions: Vec<SpecVersionResponse>,
    pub created_at: String,
    pub updated_at: String,
}

/// Schema version response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpecVersionResponse {
    pub version: String,
    pub status: String,
    #[serde(default)]
    pub schema: Option<serde_json::Value>,
}

/// Request body for the per-resource sync endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncEventTypesRequest {
    pub event_types: Vec<CreateEventTypeRequest>,
}

/// Event types resource accessor — created via [`FlowCatalystClient::event_types`].
pub struct EventTypes<'a> {
    pub(crate) client: &'a FlowCatalystClient,
}

impl EventTypes<'_> {
    /// Create a new event type.
    pub async fn create(
        &self,
        req: &CreateEventTypeRequest,
    ) -> Result<EventTypeResponse, ClientError> {
        self.client.post("/api/event-types", req).await
    }

    /// Get an event type by ID.
    pub async fn get(&self, id: &str) -> Result<EventTypeResponse, ClientError> {
        self.client.get(&format!("/api/event-types/{}", id)).await
    }

    /// Get an event type by code.
    pub async fn get_by_code(&self, code: &str) -> Result<EventTypeResponse, ClientError> {
        self.client
            .get(&format!("/api/event-types/by-code/{}", code))
            .await
    }

    /// List event types with optional filters.
    pub async fn list(
        &self,
        application: Option<&str>,
        status: Option<&str>,
        client_id: Option<&str>,
    ) -> Result<EventTypeListResponse, ClientError> {
        let mut params = Vec::new();
        if let Some(app) = application {
            params.push(format!("application={}", app));
        }
        if let Some(s) = status {
            params.push(format!("status={}", s));
        }
        if let Some(cid) = client_id {
            params.push(format!("client_id={}", cid));
        }

        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };

        self.client
            .get(&format!("/api/event-types{}", query))
            .await
    }

    /// Update an event type.
    pub async fn update(
        &self,
        id: &str,
        req: &UpdateEventTypeRequest,
    ) -> Result<EventTypeResponse, ClientError> {
        self.client
            .put(&format!("/api/event-types/{}", id), req)
            .await
    }

    /// Add a schema version to an event type.
    pub async fn add_schema_version(
        &self,
        id: &str,
        req: &AddSchemaVersionRequest,
    ) -> Result<EventTypeResponse, ClientError> {
        self.client
            .post(&format!("/api/event-types/{}/versions", id), req)
            .await
    }

    /// Archive (soft-delete) an event type.
    ///
    /// The server's DELETE on this resource is a soft archive — the row is
    /// retained with status flipped to ARCHIVED. We name it `archive` rather
    /// than `delete` to make the semantics visible.
    pub async fn archive(&self, id: &str) -> Result<(), ClientError> {
        self.client
            .delete_req(&format!("/api/event-types/{}", id))
            .await
    }

    /// Sync event types for an application — declarative reconciliation
    /// against `POST /api/applications/{appCode}/event-types/sync`.
    pub async fn sync(
        &self,
        app_code: &str,
        req: &SyncEventTypesRequest,
        remove_unlisted: bool,
    ) -> Result<crate::client::SyncResult, ClientError> {
        let query = if remove_unlisted {
            "?removeUnlisted=true"
        } else {
            ""
        };
        self.client
            .post(
                &format!("/api/applications/{}/event-types/sync{}", app_code, query),
                req,
            )
            .await
    }
}
